### PR TITLE
Fix compiled autograd.Function manual vmap rules

### LINKED ISSUE #134784: Dynamo x autograd.Function x funtorch transform: intermediate sliced tensor doesn't require grad caused incorrect result
### 🐛 Describe the bug

Let's use this small repro to demonstrate this issue:
```
import torch
import inspect

@torch.compile(backend="eager", fullgraph=True)
def fn(x):
    x = torch.vmap(gn)(x)
    return x + 1

def gn(x: torch.Tensor) -> torch.Tensor:
    if x.requires_grad:
        return x + 1
    else:
        return x - 1

x = torch.ones(2, 3, requires_grad=True)
print(fn(x))

```
The output:
```
tensor([[1., 1., 1.],
        [1., 1., 1.]], grad_fn=<AddBackward0>)
```
This means the ```gn``` function goes with the ```else``` branch, because of the intermediate sliced tensor doesn't require grad. However, if ```gn``` is an autograd function, then we would check if the input tensor requires grad and decide how to proceed.
https://github.com/pytorch/pytorch/blob/b977abd5de0ea8c37f6100e7fc170830a07ff47b/torch/_dynamo/variables/misc.py#L652-L656

We need to revisit the behavior of how to compile autograd function regarding to ```requires_grad=True/False```.

### Versions

N/A

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186330

This consolidates two reports of the same root cause: under `torch.compile`, compiled `autograd.Function` subclasses with a manual `vmap` staticmethod (`generate_vmap_rule=False`) fall off functorch's `custom_function_call` path, so the user-defined vmap rule is dropped and the compiled results/gradients disagree with eager.

Dynamo was deciding how to lower autograd.Function.apply from the
requires_grad metadata it saw on the current arguments. Under vmap, the
BatchedTensor wrapper can report requires_grad=False, so Dynamo took the
inference/direct-forward path and skipped functorch's custom_function_call
machinery. For autograd.Function subclasses with generate_vmap_rule=False,
that drops the user-defined vmap staticmethod and produces compiled results
and gradients that disagree with eager.

Mirror autograd.Function.apply's setup_context default-argument binding in
Dynamo, expose the Function subclass enough for functorch's
custom_function_call path, and route active manual-vmap cases through that
path before Dynamo chooses the autograd HOP or direct-forward lowering. Keep
plain generate_vmap_rule=True functions on the existing Dynamo path because
routing those through custom_function_call would make Dynamo trace the dynamic
class construction used by generated vmap rules.

Fixes #133493
Fixes #134784
Generated by my agent

Test Plan:
- python test/dynamo/test_autograd_function.py AutogradFunctionFunctorchTests.test_vmap_uses_custom_rule_when_batched_input_requires_grad AutogradFunctionFunctorchTests.test_vmap_custom_rule_with_unbatched_requires_grad_input AutogradFunctionFunctorchTests.test_vmap_generate_rule_compiled AutogradFunctionFunctorchTests.test_vmap_custom_rule_binds_default_args AutogradFunctionFunctorchTests.test_setup_context_binds_default_args_without_functorch AutogradFunctionFunctorchTests.test_setup_context_preserves_varargs AutogradFunctionFunctorchTests.test_old_style_autograd_function_with_grad_compiled
- python test/dynamo/test_autograd_function.py -k AutogradFunctionFunctorchTests
- python test/dynamo/test_autograd_function.py
- lintrunner -a

Benchmark Results:
Command: 25 iterations of torch.compile(..., backend="eager", fullgraph=True), torch._dynamo.reset() between iterations, measuring compile plus first backward for setup_context autograd.Function cases.
- plain_setup_context baseline: median=0.029464s mean=0.042019s; patched: median=0.027834s mean=0.087790s
- generate_vmap_rule baseline: median=0.064206s mean=0.073260s; patched: median=0.061552s mean=0.069162s

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the dynamo_wrapped CI failures were the functorch test turning into an unexpected success after the fix; I removed the stale expected-failure marker. During the rebase I also fixed the DispatchKeySet '&' handling needed by the traced functorch path, and the targeted dynamo/functorch tests plus lintrunner now pass locally.

**Comment by jansel** [comment]
My agent says the remaining dynamo_wrapped failures were unexpected successes for the output_mark_dirty functorch autograd.Function test. I removed that stale dynamo expected-failure marker as well; the exact output_mark_dirty and sibling input_mark_dirty tests pass locally with PYTORCH_TEST_WITH_DYNAMO=1, and lintrunner passes.

**Comment by jansel** [comment]
My agent says the remaining dynamo_wrapped failures were stale expected-failure markers: the affected autograd.Function functorch cases now pass, so I removed the matching CPU and CUDA Dynamo expected-failure entries. The affected CPU cases, focused Dynamo autograd.Function tests, and lintrunner pass locally; CI is running on the updated PR.

**Comment by williamwen42** [comment]
Can you rebase?

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and refreshed the implementation review. The update now guards identity- and signature-sensitive autograd.Function state, preserves wrapped/partial/custom-signature default binding, handles keyword-only and catchable binding errors consistently with eager, and keeps manual vmap dispatch ahead of positional keyword lowering. The full 84-test Dynamo autograd.Function file, focused partial/signature tests, and lintrunner pass locally; fresh CI is queued on the new head.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
